lib/std/crypto/chacha20.zig

```diff
@@ -434,3 +435,206 @@ test "crypto.chacha20 test vector 5" {
     chaCha20With64BitNonce(result[0..], input[0..], 0, key, nonce);
     testing.expectEqualSlices(u8, expected_result, result);
 }
+
+pub const chacha20poly1305_tag_size = 16;
+
+pub fn chacha20poly1305Seal(dst: []u8, plaintext: []const u8, data: []const u8, key: [32]u8, nonce: [12]u8) void {
+    assert(dst.len >= plaintext.len + chacha20poly1305_tag_size);
+
+    // derive poly1305 key
+    var polyKey = [_]u8{0} ** 32;
+    chaCha20IETF(polyKey[0..], polyKey[0..], 0, key, nonce);
+
+    // encrypt plaintext
+    chaCha20IETF(dst[0..plaintext.len], plaintext, 1, key, nonce);
+
+    // construct mac
+    var mac = Poly1305.init(polyKey[0..]);
+    mac.update(data);
+    if (data.len % 16 != 0) {
+        const zeros = [_]u8{0} ** 16;
+        const padding = 16 - (data.len % 16);
+        mac.update(zeros[0..padding]);
+    }
+    mac.update(dst[0..plaintext.len]);
+    if (plaintext.len % 16 != 0) {
+        const zeros = [_]u8{0} ** 16;
+        const padding = 16 - (plaintext.len % 16);
+        mac.update(zeros[0..padding]);
+    }
+    var lens: [16]u8 = undefined;
+    mem.writeIntSliceLittle(u64, lens[0..8], data.len);
+    mem.writeIntSliceLittle(u64, lens[8..16], plaintext.len);
+    mac.update(lens[0..]);
+    mac.final(dst[plaintext.len..]);
+}
+
+/// Verifies and decrypts an authenticated message produced by chacha20poly1305Open.
+/// Returns false if message was invalid or authentication failed.
```

Comment:
It looks like this `bool` is being used to report errors. We have a better type for that! If you make the return type `!void` then it will have an inferred error set. You can then `return error.InvalidMessage` instead of `return false`, and additionally `return error.AuthenticationFailed` instead of the second `return false`, and now callers can distinguish between types of errors.

Reply:
oh yeah 😅